libp2p/EndpointTracker.h

```diff
@@ -0,0 +1,47 @@
+// Aleth: Ethereum C++ client, tools and libraries.
+// Copyright 2019 Aleth Authors.
+// Licensed under the GNU General Public License, Version 3.
+
+#pragma once
+
+#include "Common.h"
+
+namespace dev
+{
+namespace p2p
+{
+// Class for keeping track of our external endpoint as seen by our peers.
+// Keeps track of what external endpoint is seen by every peer
+// and finds which endpoint is reported most often
+class EndpointTracker
+{
+public:
+    // Register the statement about endpoint from one of the peers.
+    // Returns number of currently kept statements in favor of _externalEndpoint
+    size_t addEndpointStatement(
+        bi::udp::endpoint const& _sourceEndpoint, bi::udp::endpoint const& _externalEndpoint);
+
+    // Find endpoint with max number of statements
+    bi::udp::endpoint bestEndpoint() const;
+
+    // Remove statements older than _timeToLive
+    void garbageCollectStatements(std::chrono::seconds const& _timeToLive);
+
+private:
+    using EndpointAndTimePoint =
+        std::pair<bi::udp::endpoint, std::chrono::steady_clock::time_point>;
+    using SourceToStatementMap = std::map<bi::udp::endpoint, EndpointAndTimePoint>;
```

Comment:
This can be unsorted map.

Reply:
Boost doesn't have a hashing function defined for `udp::endpoint` (but has `operator<`), so I used `map` to avoid defining it myself

Reply:
Can you check in the latest boost version? If not there we should report a bug.

Reply:
Not sure if it's a bug, maybe not every type should define hashing function?

Reply:
Looks like good candidate for being a key in a hash map.

Reply:
Reported https://github.com/boostorg/asio/issues/245

Reply:
Resurrecting this from the dead, just to say that I'm surprised that boostorg/asio#245 got no responses so far. We have been carrying custom hashing support for these types in `rippled` so it's been "out of sight, out of mind" for our codebase, but this really is something that boost should provide.

I'm happy to have Ripple sponsor a bounty to get the supported coded up, or to do it myself if nobody else is interested.